fairscale/nn/data_parallel/fsdp_optim_utils.py

```diff
@@ -24,19 +25,26 @@ def flatten_optim_state_dict(sd: Dict) -> Dict:
     for global_id, buffers in sd["state"].items():
         local_id = param_id_map[global_id]
         for buffer_name, p in buffers.items():
-            if torch.is_tensor(p):
+            if is_singleton_tensor(p):
+                singleton_state[local_id][buffer_name] = p
+            elif torch.is_tensor(p):
                 if buffer_name not in new_state[local_id]:
                     new_state[local_id][buffer_name] = []
                 new_state[local_id][buffer_name].append(p.reshape(-1))
+            elif isinstance(p, list):
+                singleton_state[local_id][buffer_name] = p
             else:
                 non_tensor_state[buffer_name] = p
-
     # Now combine all tensors in each buffer using torch.cat().
     for local_id, state in new_state.items():
         for buffer_name, tensors in state.items():
             new_state[local_id][buffer_name] = torch.cat(tensors)
         new_state[local_id].update(non_tensor_state)
+        new_state[local_id].update(singleton_state[local_id])
     new_sd = {"state": new_state, "param_groups": copy.deepcopy(sd["param_groups"])}
+    for k in sd.keys():  # if there are extra keys, like loss_scale, don't delete them
+        if k not in {"state", "param_groups", "uncollected_local_ids", "param_id_map"}:
```

Comment:
I'm slightly uneasy about this falling out of sync with line 160. Thoughts on some way to enforce parity between them?

Reply:
module level constant + comment + assert

```python
# These return keys are used by fairseq. To change, add @sshleifer as a reviewer.
UNFLAT_RETURN_KEYS = {"state", "param_groups", "uncollected_local_ids", "param_id_map"}
...
assert set(unflat_optim_state_dict.keys()) == UNFLAT_RETURN_KEYS 
```